---
title: "Transposable Elements (TEs) analysis in E12.5 placentae from OgtY841A-het Females x OgtY841A-homo males"
date: "June 29th, 2022"
author: 
- name: Sara Formichetti
  affiliation: EMBL Rome
  email: sara.formichetti@embl.it
output:
  prettydoc::html_pretty:
    toc: true
    toc_depth: 3
    number_sections: true
    fig_caption: yes
---

```{r Setting general options, include=FALSE}

knitr::opts_chunk$set(autodep = TRUE, cache = TRUE, echo = FALSE, warning = FALSE, message = FALSE, results = 'hide')

```

```{r loading needed packages}

library(ggplot2)
library(data.table)
library(pheatmap)
library(DESeq2)
library(gridExtra)
library(ggrepel)
library(ggpubr)

```

# Reading the data

```{r load featureCounts output tables}

# specifying the directory containing featureCounts outputs
dir <- "../../data/sequencing/featureCounts/STAR_TE"

# getting file names of featureCounts output count tables, sorted alphabetically
files <- sort(list.files(path = dir)[grep("counts.txt$", list.files(path = dir))])

# character vector of names of files with path
files_w_path <- file.path(dir, files)

# reading the count tables: first line is a commented line containing featureCounts command and needs to be skipped; columns chr,start,end,strand are dropped because they contain the comma-separated list of chr,start,end,strand for all occurrences of the TE with that unique repName in the annotation - in case of single copy genes these columns usually contain one value, in this case a list of them because each element is repeated in the genome.
counts_list <- lapply(files_w_path, function (f) {fread(f, skip = 1, drop = c("Chr", "Start",	"End", "Strand"))})
# all counts tables have the same dimensions because featureCounts counts for all meta-features (and puts 0 in case there are 0 count for that feature in that sample), therefore there will be no NAs in the reduced dt
counts_dt <- Reduce(function(...) merge(..., by = c("Geneid", "Length"), all = TRUE), counts_list)
# remove path from sample names
names(counts_dt) <- sub(".*/", "", sub("_.*", "", names(counts_dt)))
names(counts_dt)[1] <- "repName"

rm(counts_list)

```

```{r merging with TE annotation}

# loading TE annotation
TE_annotation_bed <- "../../data/annotations/RepeatMasker_RepeatLibrary20140131_mm10.noGenes.noSimple.bed"
TE_anno <- fread(TE_annotation_bed, fill = TRUE, col.names = c("seqnames", "start", "end", "repName", "strand", "repFamClass"), drop = 5)
# adding +1 to start is necessary for compatibility with methylBase-derived objects because bed file coordinates are 0-based
TE_anno$start <- TE_anno$start + 1
# making column with repFamilies only and column with repClasses only
TE_anno$repFamily <- gsub(".*/", "", TE_anno$repFamClass)
TE_anno$repClass <- gsub("/.*", "", TE_anno$repFamClass)

# merging the count table with TE annotation to get repFamilies and Classes
counts_dt <- merge(counts_dt, unique(TE_anno[,c("repName", "repClass", "repFamily", "repFamClass")]), by = "repName", all.x = TRUE, all.y = FALSE)
# removing rows whose repFamClass is NA
counts_dt <- counts_dt[!is.na(counts_dt$repFamClass)]
# removing outlier sample P1 (see quality-check heatmap in DE_analysis.Rmd) and P9 because of higher DNA FPKM
counts_dt <- counts_dt[, !(names(counts_dt) %in% c("P1", "P9")), with = FALSE]

```

The input files are:

* the output tables made by featureCounts (run in rule 'count_on_TE' of the Snakefile), which contain read counts of all repetitive elements (i.e. all repNames) for all samples
* the repetitive elements annotation `r TE_annotation_bed` (converted to SAF format in rule of Snakefile) used for featureCounts command, that is used to retrieve repFamily and repClass for each repName. The annotation I chose is the most updated library to Nov2020 of rmsk, from which I removed 'Simple repeats' and 'Low complexity regions'
* a table with STAR input reads per sample
* the samples' table, containing info on experimental design:

```{r samples table}

# loading samples table
samples_table <- fread("../../data/sequencing/samples_table.csv")
# removing outlier sample P1 (see quality-check heatmap in DE_analysis.Rmd) and P9 because of higher DNA FPKM
samples_table <- samples_table[!(samples_table$sample %in% c("P1", "P9")), ]
# reorder condition levels for better visualization and plotting
samples_table$condition <- factor(samples_table$condition, levels = c("F_HET", "M_WT", "F_HOMO", "M_HEMI"))
setkey(samples_table, sample)
# ordering counts dt based on samples table
setcolorder(counts_dt, c("repName", "repClass", "repFamily", "repFamClass", samples_table$sample))

```

# Analysis at family level 

## FPKM for each family of Repetitive Elements

For each family of Repetitive Elements (in case of elements with no repFamily name or repFamilies belonging to more than one repClass I use repClass) I compute **FPKM** values as follows: for each sample:

* I compute the sum of counts for all elements belonging to that repFamily
* I divide this sum by the total number of STAR input reads (/2 because with featureCounts I quantify fragments) for that sample and multiply by 10⁶
* I divide this number by the total sum of lengths (in Kb) of the elements belonging to that repFamily --> FPKM
* I subtract from each FPKM the total FPKM of all transposons belonging to the DNA repClass.

```{r load library size table}

# loading table containing total number of STAR input reads per sample
library_size_dt <- fread("../../data/sequencing/stats/STAR_TE/STAR_input_reads.txt")
library_size_dt$sample <- gsub(".*/", "", gsub("_.*", "", library_size_dt$V1))
library_size_dt$STAR_input <- library_size_dt$V2
library_size_dt <- library_size_dt[, .(sample, STAR_input)]
# since I counted fragments in featureCounts, I divide the STAR input reads by 2 to get a rough estimation of read pairs
library_size_dt$library_size <- library_size_dt$STAR_input/2
# ordering the library size dt by sample name
setorder(library_size_dt, sample)
# removing outlier sample from DE_analysis.Rmd and P9 because of higher DNA FPKM
library_size_dt <- library_size_dt[library_size_dt$sample %in% samples_table$sample, ]

```

```{r family FPKM}

# I notice that some elements do not have a repFamily but they have a repClass and that some repFamilies are assigned to more than one repClass, therefore I use 'repFamClass' to compute the FPKM 
# selecting columns on which to compute the sum by 'repFamClass'
cols <- c("Length", names(counts_dt)[names(counts_dt) %in% samples_table$sample])
# sum by 'repFamClass'
family_counts_dt <- counts_dt[, lapply(.SD, sum), by = .(repFamClass), .SDcols = cols] 
# remove rows with rowMeans < 5
family_counts_dt <- family_counts_dt[rowMeans(family_counts_dt[, cols[2:length(cols)], with = FALSE]) > 5, ]
# divide each column by library size and multiply by 10^6
family_FPKM_dt <- sweep(family_counts_dt[, cols[2:length(cols)], with = FALSE], 2, library_size_dt$library_size/1000000, FUN = '/')
# divide each row by repFamClass total length
family_FPKM_dt <- as.data.table(sweep(family_FPKM_dt, 1, family_counts_dt$Length/1000, FUN = '/'))
family_FPKM_dt$repFamClass <- family_counts_dt$repFamClass
# subtract DNA transposons' FPKM from each other family's FPKM value
DNA_TE_FPKM <- setNames(as.numeric(colSums(family_FPKM_dt[grep("DNA/", family_FPKM_dt$repFamClass), 1:(ncol(family_FPKM_dt)-1)])), cols[2:length(cols)])
family_FPKM_minusDNA_dt <- as.data.table(sweep(family_FPKM_dt[, 1:(ncol(family_FPKM_dt)-1)], 2, DNA_TE_FPKM, FUN = "-"))

# add back columns with repFamilies and repClasses
family_FPKM_dt$repFamily <- gsub(".*/", "", family_FPKM_dt$repFamClass)
family_FPKM_dt$repClass <- gsub("/.*", "", family_FPKM_dt$repFamClass)
family_FPKM_minusDNA_dt$repFamClass <- family_FPKM_dt$repFamClass
family_FPKM_minusDNA_dt$repFamily <- family_FPKM_dt$repFamily
family_FPKM_minusDNA_dt$repClass <- family_FPKM_dt$repClass

```

## FPKM of specific families per condition

The values are the FPKM corrected for DNA transposons.

```{r FPKM plot per condition (Supplementary Figure S4C), include = TRUE}

family_FPKM_minusDNA_dt_melt <- melt(family_FPKM_minusDNA_dt, variable.name = "sample", value.name = "corrected_FPKM")
family_FPKM_minusDNA_dt_melt <- merge(family_FPKM_minusDNA_dt_melt, samples_table, by = "sample" )
family_FPKM_minusDNA_dt_melt$sex <- gsub("_.*", "" , family_FPKM_minusDNA_dt_melt$condition)
family_FPKM_minusDNA_dt_melt$genotype <- gsub(".*_", "" , family_FPKM_minusDNA_dt_melt$condition)

plot_family_FPKM <- function (my_repFamily) {
  ggplot(family_FPKM_minusDNA_dt_melt[repFamily == my_repFamily], aes(x = condition, y = corrected_FPKM, color = sex, label = sample)) + 
  geom_point(position=position_jitter(w=0.1,h=0), size = 3) + 
  stat_summary(fun = "mean", geom = "crossbar", width = 0.5, colour = "black") +
  ylab("FPKM") +
  ggtitle(my_repFamily) +
  scale_y_log10() +
  theme(title = element_text(size = 20),
        legend.text = element_text(size = 20), legend.title = element_text(size = 20),
        axis.text = element_text(size = 19, angle = 60, vjust = 0.5, hjust = 1),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        axis.line = element_line(colour = "black"))
} 

lapply(c("L1", "ERVK", "ERVL", "Satellite"), FUN = plot_family_FPKM)

```

### testing the difference in the mean FPKM per condition

I use t-test on the FPKM values corrected for DNA transposons.

```{r FPKM diff stats, results = 'asis'}

ttest_meanFPKM_my_repFam <- function (my_repFamily) {
  p_F <- wilcox.test(family_FPKM_minusDNA_dt_melt[repFamily == my_repFamily & condition == "F_HOMO", ]$corrected_FPKM, family_FPKM_minusDNA_dt_melt[repFamily == my_repFamily & condition == "F_HET", ]$corrected_FPKM)$p.value
  p_M <- wilcox.test(family_FPKM_minusDNA_dt_melt[repFamily == my_repFamily & condition == "M_HEMI", ]$corrected_FPKM, family_FPKM_minusDNA_dt_melt[repFamily == my_repFamily & condition == "M_WT", ]$corrected_FPKM)$p.value
  paste0(my_repFamily, " - F pval = ", p_F, " - M pval = ", p_M)
}
lapply(c("L1", "ERVK", "ERVL", "Satellite"), ttest_meanFPKM_my_repFam)

```

# DE-Seq analysis of RNA transposons

I include the FPKM of DNA transposons as confounding factor in DESeq2 formula.

```{r creating DESeq inputs}

setkey(samples_table, sample)
# Selecting only RNA transposons and only columns with numeric values
RNA_TE_counts_dt <- counts_dt[!grepl("DNA|rRNA|\\?|scRNA|Other|snRNA|tRNA|Unknown|srpRNA|RC|^RNA|Satellite", counts_dt$repClass), samples_table$sample, with = FALSE]

# A count matrix will be the input for DESeq
counts_m <- as.matrix(RNA_TE_counts_dt)
rownames(counts_m) <- counts_dt[!grepl("DNA|rRNA|\\?|scRNA|Other|snRNA|tRNA|Unknown|srpRNA|RC|^RNA|Satellite", counts_dt$repClass), ]$repName
# the design df is created from the samples_table
design_df <- data.frame(condition = samples_table$condition, litter = samples_table$group_or_time_point)
# the rownames of the design df have to be the same (and in the same order) of the colnames of the count matrix
rownames(design_df) <- samples_table$sample
# adding another DNA transposons FPKM as an additional variable, that can be used as counfounding factor
design_df$DNA_FPKM <- DNA_TE_FPKM
design_df$litter <- as.character(design_df$litter)
design_df$sex <- gsub("_.*", "" , design_df$condition)

```

```{r creating DESeq Dataset}

dds <- DESeqDataSetFromMatrix(countData = counts_m,
                              colData = design_df,
                              design = ~ DNA_FPKM + condition)

```

Before running the Differential Expression analysis, the data are pre-filtered to remove all repetitive elements with < 10 reads among all samples.

```{r prefiltering}

keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep,]

```

```{r DESeq function}

dds <- DESeq(dds)

```

```{r adj pvalue cutoff}

adjpvalue_cutoff <- 0.05
l2fc_cutoff_to_label <- 0.2
baseMean_cutoff_to_label <- 20

```

```{r DESeq results list}

res_list <- list(
res_F = lfcShrink(dds, contrast=c("condition","F_HOMO","F_HET"), alpha=adjpvalue_cutoff, type="ashr"),
res_M = lfcShrink(dds, contrast=c("condition","M_HEMI","M_WT"), alpha=adjpvalue_cutoff, type="ashr")
)

```

## Differentially expressed TEs

```{r ranked DE genes, results='markup'}

# Function which takes a DESeq results object and return it as a dataframe, with genes ordered by pvalue and filtered by adjusted pvalue and baseMean
pvalue_ranked_DE <- function (my_res) {
  # keeping only genes with adjusted pvalue below my threshold
  my_res_f <- subset(my_res, padj < adjpvalue_cutoff & baseMean > 10)
  # ordering by pvalue
  my_res_f_ord <- my_res_f[order(my_res_f$pvalue),]
  return(my_res_f_ord)
}

pvalue_ranked_DE_list <- lapply(res_list, pvalue_ranked_DE)
pvalue_ranked_DE_list[[2]]

```

## MA-plots

* The threshold used for a dot to be coloured in the MA-plots is p-value adjusted < `r adjpvalue_cutoff`.
* Transposable elements whose mean expression > `r baseMean_cutoff_to_label` and abs(log2FoldChange) > `r l2fc_cutoff_to_label` are labeled.

```{r customizing plot DESeq2}

MA_plot <- function(my_dds_res, adjpval_to_color, l2fc_to_label, baseMean_to_label, ylims, xmin) {
  # adding label to genes which are significant and expressed above defined thresholds
  my_dds_res$to_label <- ""
  my_dds_res$to_label[my_dds_res$baseMean >= baseMean_to_label & my_dds_res$padj < adjpval_to_color & (my_dds_res$log2FoldChange >= l2fc_to_label | my_dds_res$log2FoldChange <= -l2fc_to_label)] <- rownames(my_dds_res[my_dds_res$baseMean >= baseMean_to_label & my_dds_res$padj < adjpval_to_color & (my_dds_res$log2FoldChange >= l2fc_to_label | my_dds_res$log2FoldChange <= -l2fc_to_label), ])
  # adding label to genes which are significant and expressed above defined thresholds
  my_dds_res$to_color <- ""
  my_dds_res$to_color[my_dds_res$padj < adjpval_to_color] <- paste0("adjpval < ", adjpval_to_color)
  dt_to_plot <- as.data.table(my_dds_res)
  ggplot(dt_to_plot, aes(x = baseMean, y = log2FoldChange)) +
    geom_point(aes(color = to_color), alpha = 0.8, size = 1.5, show.legend = FALSE) +
    geom_text_repel(aes(label = to_label), size = 2, max.overlaps = 100, show.legend = FALSE) +
    scale_x_log10(limits=c(xmin, max(my_dds_res$baseMean))) +
    ylim(ylims) +
    scale_color_manual(values = c("grey", "blue")) +
    ggtitle(gsub(".*: condition ", "", mcols(my_dds_res)$description[4])) +
    theme(axis.text = element_text(size = 10), axis.title = element_text(size = 10), 
          title = element_text(size = 10),
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          panel.background = element_blank(),
          axis.line = element_line(colour = "black"),
          panel.border = element_rect(colour = "black", fill=NA))
}

```

```{r MA plots (FIGURE), include = TRUE}

do.call(grid.arrange, lapply(res_list, function (r) {MA_plot(r, adjpval_to_color = adjpvalue_cutoff, l2fc_to_label = l2fc_cutoff_to_label, baseMean_to_label = baseMean_cutoff_to_label, xmin = 9.5, ylims = c(-2,2))}))

```

## FPKM of differentially expressed TEs

I also include two non-DE TEs, one expressed and one non-expressed in murine trophoblast cells (Ext.Fig3g in [Weigert et al. 2023](https://www.nature.com/articles/s41556-023-01114-y) ).

```{r FPKM of DE rep (FIGURE), include = TRUE}

cols <- c("Length", names(counts_dt)[names(counts_dt) %in% samples_table$sample])
# only repeats with rowMeans > 5
counts_dt_f <- counts_dt[rowMeans(counts_dt[, cols[2:length(cols)], with = FALSE]) > 5, ]
# divide each column by library size and multiply by 10^6
rep_FPKM_dt <- sweep(counts_dt_f[, cols[2:length(cols)], with = FALSE], 2, library_size_dt$library_size/1000000, FUN = '/')
# divide each row by rep length
rep_FPKM_dt <- as.data.table(sweep(rep_FPKM_dt, 1, counts_dt_f$Length/1000, FUN = '/'))
rep_FPKM_dt$repName <- counts_dt_f$repName
rep_FPKM_dt$repFamClass <- counts_dt_f$repFamClass
# long table for plotting
rep_FPKM_dt_melt <- melt(rep_FPKM_dt, variable.name = "sample", value.name = "FPKM")
rep_FPKM_dt_melt <- merge(rep_FPKM_dt_melt, samples_table[, c("sample", "condition")], by = "sample" )
rep_FPKM_dt_melt$sex <- gsub("_.*", "" , rep_FPKM_dt_melt$condition)
rep_FPKM_dt_melt$genotype <- gsub(".*_", "" , rep_FPKM_dt_melt$condition)

my_TEs <- rownames(pvalue_ranked_DE_list$res_M)
names(my_TEs) <- rownames(pvalue_ranked_DE_list$res_M)
ROIs_dt <- rep_FPKM_dt_melt[repName %in% c(my_TEs, "IAP-d-int", "IAPEz-int")]
ROIs_dt$repName <- factor(ROIs_dt$repName, levels = c(my_TEs, "IAP-d-int", "IAPEz-int"))

ggplot(data = ROIs_dt, aes(x = condition, y = FPKM, color = sex)) +
  geom_violin() +
  stat_summary(fun = "mean", geom = "crossbar", width = 0.5) +
  geom_point(position = position_jitter(width = 0.1)) +
  facet_wrap("repName", scales = "free_y", nrow = 3) +
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        axis.line = element_line(colour = "black"))

```

```{r sessioninfo, echo = TRUE, results = 'markup'}

sessionInfo()

```